### PR TITLE
Unit test archiving modified deletes both physical files

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-assets/issues/371

Versioned PR to fix issue https://github.com/silverstripe/silverstripe-versioned/pull/286

This unit test is expected to fail until the versioned PR above is merged

### CODE REVIEW

File: tests/php/FileTest.php [hunk 1]
```diff
@@ -999,4 +1001,55 @@ public function testCanEditFallsBackToCheckingDefaultFilePermissions()
...
     }
+
+    private function createModifiedFile()
+    {
+        $file = File::create();
+        $store = $this->getAssetStore();
+
+        $firstHash = 'a1c09d076de11aabc32b73ae2caca01f2b0533d9';
+        $file->setFromString('first version', 'file.txt');
+        $file->publishSingle();
```
Comment (maxime-rainville) [inline tests/php/FileTest.php:1012]:
```suggestion
        $file->setFromString('first version', 'file.txt');
        $firstHash = $file->getHash();
        $file->publishSingle();
```

File: tests/php/FileTest.php [hunk 2]
```diff
@@ -999,4 +1001,55 @@ public function testCanEditFallsBackToCheckingDefaultFilePermissions()
...
+
+        $secondHash = 'c7e7f59fa9aaed3df124d32e60982726a40568f9';
+        $file->setFromString('second version', 'file.txt');
```
Comment (maxime-rainville) [inline tests/php/FileTest.php:1015]:
```suggestion
        $file->setFromString('second version', 'file.txt');
        $secondHash = $file->getHash();
```

File: tests/php/FileTest.php [hunk 3]
```diff
@@ -999,4 +1001,114 @@ public function testCanEditFallsBackToCheckingDefaultFilePermissions()
...
+     * Validate the private helper function above works correctly
+     */
+    public function testCreateModifiedFile()
+    {
+        list($file, $store, $firstHash, $secondHash) = $this->createModifiedFile();
+
+        // Assert that our test file is a modified state with 2 different physical files
+        $this->assertTrue($file->stagesDiffer());
+        $this->assertSame('public', $store->getVisibility('file.txt', $firstHash));
+        $this->assertSame('protected', $store->getVisibility('file.txt', $secondHash));
```
Comment (maxime-rainville) [inline tests/php/FileTest.php:1033]:
```suggestion
        $this->assertSame(AssetStore::VISIBILITY_PUBLIC, $store->getVisibility('file.txt', $firstHash));
        $this->assertSame(AssetStore::VISIBILITY_PROTECTED, $store->getVisibility('file.txt', $secondHash));
```

File: tests/php/FileTest.php [hunk 4]
```diff
@@ -999,4 +1001,114 @@ public function testCanEditFallsBackToCheckingDefaultFilePermissions()
...
+     * Validate the private helper function above works correctly
+     */
+    public function testCreateModifiedFileWithDifferentFilename()
+    {
+        list($file, $store, $firstHash, $secondHash) = $this->createModifiedFileWithDifferentFilename();
+
+        // Assert that our test file is a modified state with 2 different physical files
+        $this->assertTrue($file->stagesDiffer());
+        $this->assertSame('public', $store->getVisibility('file.txt', $firstHash));
+        $this->assertSame('protected', $store->getVisibility('file-changed.txt', $secondHash));
```
Comment (maxime-rainville) [inline tests/php/FileTest.php:1064]:
```suggestion
        $this->assertSame(AssetStore::VISIBILITY_PUBLIC, $store->getVisibility('file.txt', $firstHash));
        $this->assertSame(AssetStore::VISIBILITY_PROTECTED, $store->getVisibility('file-changed.txt', $secondHash));
```

**Comment by emteknetnz** [comment]
@maxime-rainville Have added additional unit-tests as requested.

Have run this PR along with the versioned PR through [travis kitchen-sink](https://travis-ci.org/github/creative-commoners/cwp-recipe-kitchen-sink/builds/715358218), is green

Note: the docs on keep_archived_assets appear to be incorrect.  I've previously raised an issue for it https://github.com/silverstripe/silverstripe-framework/issues/9613

**Pull request merged**